infer/resource_test.go

```diff
@@ -16,15 +16,112 @@ package infer
 
 import (
 	"context"
+	"reflect"
 	"testing"
 
-	provider "github.com/pulumi/pulumi-go-provider"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/diag"
 	r "github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"pgregory.net/rapid"
+
+	provider "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer/internal/ende"
+	rRapid "github.com/pulumi/pulumi-go-provider/internal/rapid/resource"
 )
 
+// testGetDependencies runs a property test that asserts on the flow between inputs and
+// outputs for some I, O ∈ GoType and ∀ (old,new) ∈ I, out ∈ O.
+func testGetDependencies[I any, O any](t *testing.T,
+	wire func(FieldSelector, *I, *O),
+	assert func(t *testing.T, oldInput, newInput, output r.PropertyMap),
+) {
+	var i I
+	var o O
+	wireDeps := func(f FieldSelector) {
+		if wire != nil {
+			wire(f, &i, &o)
+		}
+	}
+	setDeps, err := getDependenciesRaw(
+		&i, &o, wireDeps,
+		false, /*isCreate*/
+		true /*isPreview*/)
+	require.NoError(t, err)
+
+	inputT := rapid.Just(reflect.TypeOf(i))
+	outputT := rapid.Just(reflect.TypeOf(o))
+
+	getMap := func(t rRapid.Typed) r.PropertyMap {
+		return t.Value.ObjectValue()
+	}
+
+	rapid.Check(t, func(rt *rapid.T) {
+		oldInput := rapid.Map(rRapid.ValueOf(inputT), getMap).
+			Draw(rt, "oldInput")
+		newInput := rapid.Map(rRapid.ValueOf(inputT), getMap).
+			Draw(rt, "newInput")
+		output := rapid.Map(rRapid.ValueOf(outputT), getMap).
+			Draw(rt, "output")
+
+		setDeps(oldInput, newInput, output)
+
+		assert(t, oldInput, newInput, output)
+	})
+}
+
+func TestDefaultDependencies(t *testing.T) {
+	t.Parallel()
+	type input struct {
+		I1 string            `pulumi:"i1"`
+		I2 *int              `pulumi:"i2,optional"`
+		I3 map[string]string `pulumi:"i3"`
+	}
+
+	type output struct {
+		input
+
+		O1 *string        `pulumi:"o1,optional"`
+		O2 float64        `pulumi:"o2"`
+		O3 map[string]int `pulumi:"o2"`
+	}
+
+	assert := func(t *testing.T, oldInput, newInput, output r.PropertyMap) {
+		if newInput.ContainsUnknowns() {
+			for k, v := range output {
+				if newV, ok := newInput[k]; ok &&
+					ende.DeepEquals(newV, v) {
+					continue
+				}
+				assert.True(t, ende.IsComputed(v),
+					"key: %q", string(k))
+			}
+		} else if !ende.DeepEquals(
+			r.NewObjectProperty(oldInput),
+			r.NewObjectProperty(newInput)) {
+			// If there is a change, then every item item should be
```

Comment:
Hmm, why is this? This is simulating preview only I guess but still?  It's a little interesting, trying to wrap my head around. So the implementation is allowed to copy oldInput to output, or mark outputs computed.

Reply:
The default logic is: "if any input property has changed, declare all the output properties as unknown computed, except the outputs that are just pass-through inputs". We mark all output properties as unknown because we don't know if the changed input will change the output. We do know if passthrough inputs are known.

Reply:
In this condition:

```go
if n, ok := newInput[k]; !ok || ende.IsComputed(n) {
```

If `newInput[k]` makes ok==false, then there are no asserts performed at all. 

So I don't think the property you state in tested here. But implementation might still do it right! It's just not tested seemingly, or I don't get something.

Consider oldInput={"a": 1}, newInput={"a": 42}, output: {"a": 42, "b": "foo"}.

This is trying to simulate "a" input changing, I think this will pass the check although "b": "foo" didn't get marked unknown. By your logic above it seems that "b" should be unknown since we can't assume that pushing changed inputs through the resource black box won't change the outputs.

Reply:
I think you're right that the test is missing something. https://github.com/pulumi/pulumi-go-provider/pull/123/commits/c8c23a07d70f783a5048af86d9c5d5b5fcd0ec24 strengthens the test to assert our expected conditions: 

https://github.com/pulumi/pulumi-go-provider/blob/c8c23a07d70f783a5048af86d9c5d5b5fcd0ec24/infer/resource_test.go#L99-L113